#### 1. Pick one of the datasets from the ChatBot session(s) of the TUT demo (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values

I will choose to use the following dataset as given on the homework page:

"https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

After running this code, we can see that there is missing data. A zero indicates that there is not missing data for the related attribute. We can see there is missing data for "id" and "song".

#### 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a pandas DataFrame has, and then

use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,

write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset

The ChatBot session left me with the following code (Note: The ChatBot has renamed the DataFrame to "data". This is consistent going forward).

In [7]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
data = pd.read_csv(url)

# Get the number of rows and columns
rows, columns = data.shape

print(f'The dataset has {rows} rows and {columns} columns.')

The dataset has 391 rows and 11 columns.


Running this shows that the dataset has 391 rows and 11 columns however the chatbot has stated that there are 442 rows instead.

It seems based on the information the ChatBot has provided me, an observation is another word for a row. An observation contains all the data about an individual data point, in this case, a villager. An observation corresponds to its own data point and it can contain different details attributes of the data such as the Name or Species of a villager in this case.

A variable is another word for a column. It holds the attributes or details of a data point such as Name or Species in the case of a villager.

An observation contains all the information about an individual data point while a variable contains a specific attribute of a data point.

#### 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset

Below is the provided code from the ChatBot session. This code has been split up so jupyterhub can properly display it.

In [6]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
data = pd.read_csv(url)

# Basic summary of numerical columns
data.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [8]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
data = pd.read_csv(url)

# Summary of all columns (numerical and categorical)
data.describe(include='all')

row_n       id     name gender species birthday personality  \
count   391.000000      390      391    391     391      391         391   
unique         NaN      390      391      2      35      361           8   
top            NaN  admiral  Admiral   male     cat     1-27        lazy   
freq           NaN        1        1    204      23        2          60   
mean    239.902813      NaN      NaN    NaN     NaN      NaN         NaN   
std     140.702672      NaN      NaN    NaN     NaN      NaN         NaN   
min       2.000000      NaN      NaN    NaN     NaN      NaN         NaN   
25%     117.500000      NaN      NaN    NaN     NaN      NaN         NaN   
50%     240.000000      NaN      NaN    NaN     NaN      NaN         NaN   
75%     363.500000      NaN      NaN    NaN     NaN      NaN         NaN   
max     483.000000      NaN      NaN    NaN     NaN      NaN         NaN   

                song   phrase           full_id  \
count            380      391               391   
unique            92      388               391   
top     K.K. Country  wee one  villager-admiral   
freq              10        2                 1   
mean             NaN      NaN               NaN   
std              NaN      NaN               NaN   
min              NaN      NaN               NaN   
25%              NaN      NaN               NaN   
50%              NaN      NaN               NaN   
75%              NaN      NaN               NaN   
max              NaN      NaN               NaN   

                                                      url  
count                                                 391  
unique                                                391  
top     https://villagerdb.com/images/villagers/thumb/...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

In [9]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
data = pd.read_csv(url)

# Overview of the DataFrame
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB


In [10]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
data = pd.read_csv(url)

# Value counts for a categorical column, e.g., "species"
data['species'].value_counts()

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

In [11]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
data = pd.read_csv(url)

# Check for missing data
data.isnull().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [13]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
data = pd.read_csv(url)

# Basic summary of numerical columns
data.shape

(391, 11)

#### 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by df.shape and what is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column

Looking at the cells where we ran data.shape and data.describe(include='all'), we can see that each attribute in the count row should add up to 391 as there are 391 total rows in the dataset. Almost every variable has a count of 391, however, we can notice that the column for songs and id are not 391. Instead, song has a total count of 380 and id has a total count of 390. This is the discrepancy.

This can be explained by data.isnull().sum() which tells us what data is missing with respect to each variable as well as how many instances theer are of missing data. We can see that there are 11 instances of missing data for song which we can see is true as 380 is 11 less than 391. The same is true for the count of id which is 1 less than 391.

This discrepancy is due to missing data in certain variables.

#### 5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

The difference between an attribute such as data.shape and a method such as data.describe() is attributes are simple characteristics you can run that hold a value or otherwise stores information. It doesn't require any input and doesn't do any computations.

In contrast, a method does do computations and it can be given an input, or arguments. Methods execute code in the module they are imported from. Attributes do not execute any code and just store information. That is the crucial difference.

Below is the summary compiled and created by the ChatBot of our session in relation to this part of the homework. I have also included a public link to the chat log.

#### Session Summary
Here's a summary of our session today:

Dataset Exploration:

You started by providing a dataset URL, and we discussed how to load it into a pandas DataFrame.
We explored how to find the number of rows and columns in your dataset using data.shape. It was mentioned that the dataset should have 391 rows and 11 columns (based on your execution).
Key Concepts:

Observations vs. Variables:
An observation is a row in the dataset (representing a villager), and a variable is a column (representing characteristics like species or personality).
Attributes vs. Methods:
An attribute (e.g., data.shape) returns stored information about the object.
A method (e.g., data.describe()) performs some operation or calculation.
Summarizing the Dataset:

You learned how to summarize the dataset using functions like data.describe(), data.info(), and value_counts() for categorical columns.
Methods to check for missing values in the dataset, including .isnull().sum() and .info() were also discussed.
Understanding DataFrame Limits:

We touched on the practical limits of pandas DataFrames in terms of rows and columns, with the main constraint being your system's available memory.
Throughout the session, we focused on methods to understand and summarize your dataset efficiently using pandas in Python.

#### Link: https://chatgpt.com/share/7a19f795-c9ce-4763-9852-0463fe249cd1

#### 6. The df.describe() method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics

In [8]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
df = pd.read_csv(url)

# Basic summary of numerical columns
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [9]:
df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

Above I have run df.describe() again as well as the full dataframe.

After some conversation with the ChatBot, it is clear that the 'count' statistic displays the total number of numerical values that are in row_n. 
The 'mean' statistic displays the mean or the average of the numerical values in row_n. It sums all values in row_n together then divides them by 'n' number of total terms, in this case being 391. 
The 'std' statistic calculates the standard deviation of the numeric values by square rooting the sum of the squared difference of each term and the mean, then dividing it by the total number of terms 'n'-1.
The 'min' statistic displays the minimum value found in the dataframe which in this case is 2.
The '25%' statistic displays the 25 percentile value or in other words, the value that marks the greatest of the bottom 25% of values after having all the values ordered from least to greatest.
The '50%' statistic displays the median, which is the middle value of the data when ordered from least to greatest. It shows the half-way point of the data.
The '75%' statistic shows the 75 percentile value, which is the value that appears at the 75% mark when the data is ordered from least to greatest. It tells us that this value is the greatest of the bottom 75% if values.
The 'max' statistic shows the greatest numerical value in the dataframe, in this case being 483.

df.describe() does not count non-numeric or missing values. That is why here it only shows data from row_n which is the only column that has traditional numeric values (birthday also has numeric values but those pertain to dates rather than numbers so they aren't considered 'numeric values').

#### 7. Missing data can be considered "across rows" or "down columns". Consider how df.dropna() or del df['col'] should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

1. Provide an example of a "use case" in which using df.dropna() might be preferred over using del df['col']


2. Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna()


3. Discuss why applying del df['col'] before df.dropna() when both are used together could be important


3. Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.

1.

    df.dropna() would be preferred when you want to drop multiple different rows or columns of data. You don't need to individually specify which rows or columns to remove. It also has much more functionality such as changing the inplace value to true or false. Changing it to true will edit the dataframe permanently while keeping it at false will only show you the result of the changes for that once instance which may be more preferred if you do not want to directly edit the data. del df['col'] will completely delete any row and will always permanently delete the data as it lacks much functionality.
    
2.
    
    However, del df['col'] may be more preferred as it can delete ANY row or column, not only ones where there are missing values. For example, if you would like to remove irrelevant or non-numerical rows or columns that do not have any missing values, df.dropna() will not do anything.

3.

    Using both del df['col'] and df.dropna() can be helpful and important to remove unwanted data. For example, let's say there are rows and columns that have missing data. Let's also say there are rows and columns with unimportant or irrelevant data. Using df.dropna() to remove rows and columns with missing data first, then using del df['col'] to eliminate the rest of the unwanted data will be more efficient and quicker than only using one of the pieces of code. This is because df.dropna() has the capabilities to remove multiple rows and columns at once. Using df.dropna() can be beneficial before deleting columns as you can test to see how your data changes by leaving inplace as false. Using both may also be necessary as df.dropna() can only remove rows and columns with missing values and del df['col'] can only remove columns by individually specifying them.
    Both pieces of code have essential functionality for efficient and quick removal of data.

In [10]:
df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

Above once again is the dataframe, although it is too large to display here. Moving forward for question 4, I will be using df.dropna() to remove all columns and rows with missing data since removing missing data is the only task here. It makes sense to only use df.dropna() as we don't need to remove anything else. del df['col'] would be unnecessary and inefficient here. 

I will not be giving df.dropna() and inplace argument true as I do not wish to truly remove data from the dataframe and I do not feel it's necessary as all we need to do is show the before and after.

In [19]:
df.dropna(subset='id').dropna(axis=1)

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

       phrase            full_id  \
0     aye aye   villager-admiral   
1    sidekick   villager-agent-s   
2     snuffle     villager-agnes   
3     Ayyeeee        villager-al   
4    it'sa me   villager-alfonso   
..        ...                ...   
386    hay-OK    villager-winnie   
387   snarrrl  villager-wolfgang   
388   tsk tsk      villager-yuka   
389     pronk      villager-zell   
390     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[390 rows x 10 columns]

As we can see, 2 columns have been removed. We can see that 'id' and 'song' were the only 2 columns to have missing data below.

In [15]:
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

By combining methods we can see that these two columns have been removed.

In [20]:
df.dropna(subset='id').dropna(axis=1).isna().sum()

row_n          0
id             0
name           0
gender         0
species        0
birthday       0
personality    0
phrase         0
full_id        0
url            0
dtype: int64

I believe this to be most efficient way to remove unncessary data while keeping others as there is only one value missing for id so I believe it would be best to remove the row as to not discard the rest of the values in id. I've decided to remove the full column for song as there are many values for song that are missing. This way we don't need to unnecessarily get rid of a bunch of data.

Here is the linked ChatBot session which includes the summary.
https://chatgpt.com/share/2545f717-13b2-413a-9db5-319415212e75

#### 8. Give brief explanations in your own words for any requested answers to the questions below

1. Use your ChatBot session to understand what df.groupby("col1")["col2"].describe() does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you

Below is the titanic dataset.

In [21]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

According to the ChatBot, grouby will take 2 columns and compare them. The statistics shown are how many times the attribute in 'col1' appears in the data and what's the most common value it has in 'col2'. This gives us the information of how 'col1' relates to 'col2'.

In [22]:
df.groupby('survived')['class'].describe()

count unique    top freq
survived                         
0          549      3  Third  372
1          342      3  First  136

Here we can see the comparison of whether someone survived and what class they are in. For the survived column, 0 means false and 1 means true. We can see from this dataframe that most of the people that did not survive were in third class.

2. Assuming you've not yet removed missing values in the manner of question "7" above, df.describe() would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like df.groupby("col1")["col2"].describe()?

In [25]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

df.describe() is fundamentally different then df.groupby("col1")["col2"].describe() in terms of the 'count' value as df.describe() gives the unsorted unaltered data from the dataframe while df.groupby("col1")["col2"].describe() doesn't give data, rather it gives statistics about the data. It compares 2 columns and peforms some calculations about them. In this case, it groups the column 'survived' by each of it's different values, 1 and 0. Which will give it a different count showing how many times a 1 or 0 appears while the count value from df.describe() is the total number of data values in the survived column. df.groupby("col1")["col2"].describe() will split the total count value up to provide information.

3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT

A. Forget to include import pandas as pd in your code

When asking the ChatBot for assistance, it gives me immediate feedback and a correct solution that I can use to solve the problem. When using Google, I find that I need to go through a website and read a discussion forum to create an answer. Although this still isn't as easy as using a ChatBot as the ChatBot can apply the context of previous chats to the solution and give me a personalized solution that is easy for me to fix and understand.

B. Mistype "titanic.csv" as "titanics.csv"

After asking Google and the ChatBot for assistance, I found that the ChatBot again instantly recognized the issue from previous context and gave me a simple and personalized solution. I found that this time, Google was giving me wildly different answers and threads, none of which were to spell the file name correctly.

I have also tried removing a 't' from 'githubusercontent' in the url. The ChatBot gave me various solutions that did not really work however eventually when asking enough, it gave me a solution to my problem that worked flawlessly. Once again, Google didn't really provide much assistance as it lacked the context of my code and my goal. It also wasn't very personalized and included other more confusing info as it gave me links to forum websites where people have very different codes and therefore required very different solutions.

C. Try to use a dataframe before it's been assigned into the variable

You can simulate this by just misnaming the variable. For example, if you should write `df.groupby("col1")["col2"].describe()` based on how you loaded the data, then instead write `DF.groupby("col1")["col2"].describe()`

Along with once again giving me a very easy solution while also recognizing the problem quickly, this time ChatBot also offered some explanation about the code to teach me why things were not working. It had told me that code is case sensitive and that DF does not mean the samer thing as df. Google however is not case sensitive so it did not even pick up on how DF in the error was capitalized unlike the ChatBot.

D. Forget one of the parentheses somewhere the code

Once again, ChatBot told me my mistake and how to fix it as well as giving me the correct syntax for df.groupby()[]. Google also gave me a simple valid solution, however the ChatBot not only provides me with a personalized solution but also gives me an opportunity to learn and gain a deeper understanding of my mistakes. Google does not help with that much at all.

E. Mistype one of the names of the chained functions with the code

F. Use a column name that's not in your data for the groupby and column selection

G. Forget to put the column name as a string in quotes for the groupby and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question

Once again, for all of these, the ChatBot gave me personalized solutions with extra information to help me understand and fix my code as well as telling me why I made those mistakes and how to avoid them. Google varied with instructions and methods to resolve the issue, often times the solutions not working. Google created a much more time consuming and frustrating experience while also not teaching me anything about python or my code in general.

#### 9. Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

Yes, I have. At first I was skeptical about the use of a ChatBot and it seemed like more of a hassle than a tool to use. I was very against it however having worked on this homework, I've begun to appreciate what the ChatBot can offer and look forward to using it in the future. I believe this ChatBot will be my greatest asset in this course as well as my TA Pranav. They were both very helpful in giving me advice or explaining something in a deeper manner.